drivers/gpio/gpio_ambiq.c

```diff
@@ -0,0 +1,359 @@
+				    pincfg.GP.cfg_b.eGPOutCfg == AM_HAL_GPIO_PIN_OUTCFG_OPENDRAIN) {
+					op |= BIT(i);
+				}
+			}
+		}
+		*outputs = op;
+	}
+
+	return 0;
+}
+#endif
+
+static int ambiq_gpio_port_get_raw(const struct device *dev, gpio_port_value_t *value)
+{
+	const struct ambiq_gpio_config *const dev_cfg = dev->config;
+
+	*value = (*AM_HAL_GPIO_RDn(dev_cfg->offset >> 2));
+
+	return 0;
+}
+
+static int ambiq_gpio_port_set_masked_raw(const struct device *dev, gpio_port_pins_t mask,
+					  gpio_port_value_t value)
+{
+	const struct ambiq_gpio_config *const dev_cfg = dev->config;
+	uint32_t pin_offset = dev_cfg->offset >> 2;
+
+	for (int i = 0; i < dev_cfg->ngpios; i++) {
+		if ((mask >> i) & 1) {
+			am_hal_gpio_state_write(i + pin_offset, ((value >> i) & 1));
+		}
+	}
+
+	return 0;
+}
+
+static int ambiq_gpio_port_set_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	const struct ambiq_gpio_config *const dev_cfg = dev->config;
+	uint32_t pin_offset = dev_cfg->offset >> 2;
+
+	for (int i = 0; i < dev_cfg->ngpios; i++) {
+		if ((pins >> i) & 1) {
+			am_hal_gpio_state_write(i + pin_offset, AM_HAL_GPIO_OUTPUT_SET);
+		}
+	}
+
+	return 0;
+}
+
+static int ambiq_gpio_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	const struct ambiq_gpio_config *const dev_cfg = dev->config;
+	uint32_t pin_offset = dev_cfg->offset >> 2;
+
+	for (int i = 0; i < dev_cfg->ngpios; i++) {
+		if ((pins >> i) & 1) {
+			am_hal_gpio_state_write(i + pin_offset, AM_HAL_GPIO_OUTPUT_CLEAR);
+		}
+	}
+
+	return 0;
+}
+
+static int ambiq_gpio_port_toggle_bits(const struct device *dev, gpio_port_pins_t pins)
+{
+	const struct ambiq_gpio_config *const dev_cfg = dev->config;
+	uint32_t pin_offset = dev_cfg->offset >> 2;
+
+	for (int i = 0; i < dev_cfg->ngpios; i++) {
+		if ((pins >> i) & 1) {
+			am_hal_gpio_state_write(i + pin_offset, AM_HAL_GPIO_OUTPUT_TOGGLE);
+		}
+	}
+
+	return 0;
+}
+
+static void ambiq_gpio_isr(const struct device *dev)
+{
+	struct ambiq_gpio_data *const data = dev->data;
+	const struct ambiq_gpio_config *const dev_cfg = dev->config;
+
+	uint32_t int_status;
+
+	am_hal_gpio_interrupt_irq_status_get(dev_cfg->irq_num, false, &int_status);
+	am_hal_gpio_interrupt_irq_clear(dev_cfg->irq_num, int_status);
+
+	gpio_fire_callbacks(&data->cb, dev, int_status);
+}
+
+static int ambiq_gpio_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+					      enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	const struct ambiq_gpio_config *const dev_cfg = dev->config;
+	struct ambiq_gpio_data *const data = dev->data;
+
+	am_hal_gpio_pincfg_t pincfg;
+	int gpio_pin = pin + (dev_cfg->offset >> 2);
+	uint32_t int_status;
+	int ret;
+
+	ret = am_hal_gpio_pinconfig_get(gpio_pin, &pincfg);
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		pincfg.GP.cfg_b.eIntDir = AM_HAL_GPIO_PIN_INTDIR_NONE;
+		ret = am_hal_gpio_pinconfig(gpio_pin, pincfg);
+
+		irq_disable(dev_cfg->irq_num);
```

Comment:
Won't this disable IRQs for _all_ pins?

Reply:
Thanks for catching this. Yes, we cannot call irq_disable() for the GPIO pin interrupt disablement here since every 32 pins share the same irq number.